src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java

```diff
@@ -398,34 +398,48 @@ public synchronized void close() {
                         } catch (SQLException ex) {
                             logger.debug(String.format("Failed to flush before dropping entity '%s'", me.getValue().getEntity().getName()), ex);
                         } finally {
-                            if (insert != null) {
-                                try {
-                                    insert.close();
-                                } catch (Exception e) {
-                                    logger.trace("Could not close prepared statement.", e);
-                                }
-                            }
-
-                            if (insertReturning != null) {
-                                try {
-                                    insertReturning.close();
-                                } catch (Exception e) {
-                                    logger.trace("Could not close prepared statement.", e);
-                                }
-                            }
+                            closeMappedEntity(me.getValue());
                         }
 
                         dropEntity(me.getValue().getEntity());
                     } catch (DatabaseEngineException ex) {
                         logger.debug(String.format("Failed to drop entity '%s'", me.getValue().getEntity().getName()), ex);
                     }
                 }
+
+            } else {
+                entities.forEach((key, mappedEntity) -> closeMappedEntity(mappedEntity));
+            }
+
+            for (PreparedStatementCapsule preparedStatement : stmts.values()) {
```

Comment:
Can you clarify when is it appropriate to include "final"?

- You're suggesting to include "final" in the for loop, which I totally understand. However, the loop above, which was already there, does not include the "final". Should I add "final" there as well?
- What about the catch clauses? Throughout the code there are catch clauses which include the "final" and other which don't. What should I do in those cases?

Reply:
Appropriate is basically in every variable declaration that you know will never change (for primitive types, their value will never change; for objects, their reference will never change).

As a rule of thumb, you should put them in all situations that I stated. There are a lot of those missing throughout the code because we are not enforcing this rule (but mostly because it's older code). But for now, just add those to your code only. Also, Intellij inspection should highlight where "final"s are missing.

Reply:
Makes sense to me. Thank you for clarifying.